packages/router/src/lib/routes.ts

```diff
@@ -14,145 +14,169 @@ const FILES = import.meta.glob<RouteExport>([
-                  },
-                ]),
-            };
-
-            routes?.[insert](newRoute as Route);
-            return parent;
-          }
-        }
+export function createRoutes(files: Files): Route[] {
+  const rawRoutesByLevelMap = Object.keys(files).reduce((acc, filename) => {
+    const rawPath = toRawPath(filename);
+    const rawSegments = rawPath.split('/');
+    const level = rawSegments.length - 1;
+    const rawSegment = rawSegments[level];
+    const ancestorRawSegments = rawSegments.slice(0, level);
+
+    return {
+      ...acc,
+      [level]: {
+        ...acc[level],
+        [rawPath]: {
+          filename,
+          rawSegment,
+          ancestorRawSegments,
+          segment: toSegment(rawSegment),
+          level,
+          children: [],
+        },
+      },
+    };
+  }, {} as Record<number, Record<string, RawRoute>>);
+
+  const allLevels = Object.keys(rawRoutesByLevelMap).map(Number);
+  const maxLevel = Math.max(...allLevels);
+
+  for (let level = maxLevel; level > 0; level--) {
+    const rawRoutesMap = rawRoutesByLevelMap[level];
+    const rawPaths = Object.keys(rawRoutesMap);
+
+    for (const rawPath of rawPaths) {
+      const rawRoute = rawRoutesMap[rawPath];
+      const parentRawPath = rawRoute.ancestorRawSegments.join('/');
+      const parentRawSegmentIndex = rawRoute.ancestorRawSegments.length - 1;
+      const parentRawSegment =
+        rawRoute.ancestorRawSegments[parentRawSegmentIndex];
+
+      rawRoutesByLevelMap[level - 1] ||= {};
+      rawRoutesByLevelMap[level - 1][parentRawPath] ||= {
+        filename: null,
+        rawSegment: parentRawSegment,
+        ancestorRawSegments: rawRoute.ancestorRawSegments.slice(
+          0,
+          parentRawSegmentIndex
+        ),
+        segment: toSegment(parentRawSegment),
+        level: level - 1,
+        children: [],
+      };
+
+      rawRoutesByLevelMap[level - 1][parentRawPath].children.push(rawRoute);
+    }
+  }
 
-        if (root || node) {
-          const current = root ? routes : parent._children;
-          const found = current?.find((route: any) => route.path === path);
-
-          if (found) {
-            if (!found._children) {
-              found._children = [];
-            }
-
-            found.pathMatch = pathMatch;
-          } else {
-            current?.[insert]({
-              path,
-              pathMatch,
-              _module: () => module(),
-              loadChildren: () =>
-                module().then((m) => [
-                  {
-                    path: '',
-                    component: m.default,
-                    ...toRouteConfig(m.routeMeta as RouteMeta | undefined),
-                  },
-                ]),
-            });
-          }
-
-          return (
-            found ||
-            (current?.[insert === 'unshift' ? 0 : current.length - 1] as Route)
-          );
-        }
+  const rootRawRoutesMap = rawRoutesByLevelMap[0] || {};
```

Comment:
Under what scenario would this array be empty? If no routes were found?

Reply:
Yes, if `files` is an empty object. I can make this check more explicit and move it to the top of the method:

```ts
function createRoutes(files: Files): Route[] {
  const filenames = Object.keys(files);

  if (filenames.length === 0) {
    return [];
  }

  // ...
}
```

what do you think? 👀

Reply:
Works for me 👍